pkg/core/resources/apis/workload/generate/generator.go

```diff
@@ -0,0 +1,202 @@
+	"github.com/kumahq/kuma/v2/pkg/plugins/runtime/k8s/metadata"
+	mesh_cache "github.com/kumahq/kuma/v2/pkg/xds/cache/mesh"
+	xds_context "github.com/kumahq/kuma/v2/pkg/xds/context"
+)
+
+const (
+	managedByValue = "workload-generator"
+)
+
+// Generator generates Workload objects from Dataplane resources created on
+// Universal.
+type Generator struct {
+	logger              logr.Logger
+	generateInterval    time.Duration
+	deletionGracePeriod time.Duration
+	metric              prometheus.Summary
+	resManager          manager.ResourceManager
+	meshCache           *mesh_cache.Cache
+	zone                string
+}
+
+var _ component.Component = &Generator{}
+
+func New(
+	logger logr.Logger,
+	generateInterval time.Duration,
+	deletionGracePeriod time.Duration,
+	metrics core_metrics.Metrics,
+	resManager manager.ResourceManager,
+	meshCache *mesh_cache.Cache,
+	zone string,
+) (*Generator, error) {
+	metric := prometheus.NewSummary(prometheus.SummaryOpts{
+		Name:       "component_workload_generator",
+		Help:       "Summary of Workload generation duration",
+		Objectives: core_metrics.DefaultObjectives,
+	})
+	if err := metrics.Register(metric); err != nil {
+		return nil, err
+	}
+	return &Generator{
+		logger:              logger,
+		generateInterval:    generateInterval,
+		deletionGracePeriod: deletionGracePeriod,
+		metric:              metric,
+		resManager:          resManager,
+		meshCache:           meshCache,
+		zone:                zone,
+	}, nil
+}
+
+// workloadForDataplane extracts and validates the workload identifier from a dataplane
+func (g *Generator) workloadForDataplane(dataplane *core_mesh.DataplaneResource) (string, bool) {
+	workloadName, ok := dataplane.GetMeta().GetLabels()[metadata.KumaWorkload]
+	if !ok || workloadName == "" {
+		return "", false
+	}
+	allErrs := apimachineryvalidation.NameIsDNS1035Label(workloadName, false)
+	if len(allErrs) != 0 {
+		g.logger.Info("couldn't generate Workload from kuma.io/workload label, contains invalid characters", "value", workloadName, "error", allErrs, "dataplane", dataplane.GetMeta().GetName())
+		return "", false
+	}
+	return workloadName, true
+}
+
+func (g *Generator) generate(ctx context.Context, mesh string, dataplanes []*core_mesh.DataplaneResource, workloads []*workload_api.WorkloadResource) {
+	log := g.logger.WithValues("mesh", mesh)
+	workloadsByName := map[string]bool{}
+	for _, dataplane := range core_mesh.SortDataplanes(dataplanes) {
+		if workloadName, ok := g.workloadForDataplane(dataplane); ok {
+			log.V(1).Info("will generate Workload for dataplane", "dataplane", dataplane.GetMeta().GetName(), "workload", workloadName)
+			workloadsByName[workloadName] = true
+		}
+	}
+	log.V(1).Info("processing workloads", "workloadCount", len(workloadsByName), "dataplaneCount", len(dataplanes))
+
+	for _, workload := range workloads {
+		if managedBy, ok := workload.GetMeta().GetLabels()[mesh_proto.ManagedByLabel]; !ok || managedBy != managedByValue {
```

Comment:
why are we doing this? there is no flow when user would need to create Workload manually